File: data/src/androidTest/java/com/github/anrimian/musicplayer/data/database/dao/play_list/PlayListsDaoWrapperTest.java
package com.github.anrimian.musicplayer.data.database.dao.play_list;

import static com.github.anrimian.musicplayer.data.utils.TestDataProvider.composition;
import static com.github.anrimian.musicplayer.domain.utils.ListUtils.asList;
import static java.util.Collections.emptyList;

import android.content.Context;
import android.util.Log;

import androidx.core.util.Pair;
import androidx.room.Room;
import androidx.test.platform.app.InstrumentationRegistry;

import com.github.anrimian.musicplayer.data.database.AppDatabase;
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao;
import com.github.anrimian.musicplayer.data.database.entities.playlist.PlayListEntryEntity;
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlayList;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayListItem;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Date;
import java.util.List;

public class PlayListsDaoWrapperTest {

    private PlayListDao playListDao;
    private CompositionsDao compositionsDao;
    private AppDatabase db;

    private PlayListsDaoWrapper daoWrapper;

    @BeforeEach
    public void setUp() {
        Context context = InstrumentationRegistry.getInstrumentation().getContext();
        db = Room.inMemoryDatabaseBuilder(context, AppDatabase.class).build();
        compositionsDao = db.compositionsDao();
        playListDao = db.playListDao();

        daoWrapper = new PlayListsDaoWrapper(playListDao, compositionsDao, db);
    }

    @AfterEach
    public void tearDown() {
        db.close();
    }

    @Test
    public void testModifyChangeWithDuplicates() {
        StoragePlayList playList1 = new StoragePlayList(1L,
                "test",
                new Date(),
                new Date());
        StoragePlayList playList2 = new StoragePlayList(2L,
                "test1",
                new Date(),
                new Date());
        daoWrapper.applyChanges(
                asList(new Pair<>(playList1, emptyList()), new Pair<>(playList2, emptyList())),
                emptyList(),
                emptyList()
        );
        StoragePlayList duplicatePlayList = new StoragePlayList(2L,
                "test",
                new Date(),
                new Date());
//        daoWrapper.applyChanges(emptyList(), asList(new Change<>(playList1, duplicatePlayList)));
    }

    @Test
    public void testMoveItems() {
        long playlistId = daoWrapper.insertPlayList("playlist", new Date(), new Date(), () -> null);

        for (int i = 0; i < 10; i++) {
            long id = compositionsDao.insert(composition(null, null, String.valueOf(i)));
            playListDao.insertPlayListEntries(asList(new PlayListEntryEntity(
                    null,
                    id,
                    playlistId,
                    i
            )));
        }

        List<PlayListItem> items = daoWrapper.getPlayListItemsObservable(playlistId, false, null)
                .blockingFirst();
        displayItems("testMoveItems, items: ", items);

        daoWrapper.moveItems(playlistId, 0, 7);

        items = daoWrapper.getPlayListItemsObservable(playlistId, false, null)
                .blockingFirst();
        displayItems("testMoveItems, moved items: ", items);

    }

    @Test
    public void testUpdatePlaylistNameThatAlreadyExists() {
        Date date = new Date();
        StoragePlayList playList1 = new StoragePlayList(1L, "test", date, date);
        StoragePlayList playList2 = new StoragePlayList(2L, "test1", date, date);
        StoragePlayList playList3 = new StoragePlayList(3L, "test2", date, date);
        daoWrapper.insertPlayList(playList1);
        daoWrapper.insertPlayList(playList2);
        daoWrapper.insertPlayList(playList3);

        StoragePlayList duplicatePlayList2 = new StoragePlayList(2L, "test", date, date);
        StoragePlayList duplicatePlayList3 = new StoragePlayList(3L, "test", date, date);
//        daoWrapper.applyChanges(emptyList(), asList(
//                new Change<>(playList2, duplicatePlayList2),
//                new Change<>(playList3, duplicatePlayList3)
//        ));

        System.out.println("KEKAS" + daoWrapper.getPlayListsObservable().blockingFirst());
    }

    @Test
    public void testInsertPlaylistThatAlreadyExistsButWithoutStorageId() {
        Date date = new Date();
        daoWrapper.insertPlayList("test", date, date, () -> null);

        StoragePlayList playList = new StoragePlayList(1L, "test", date, date);
        daoWrapper.applyChanges(asList(new Pair<>(playList, emptyList())), emptyList(), emptyList());

        //got 2 playlists
        System.out.println("KEKAS" + daoWrapper.getPlayListsObservable().blockingFirst());
    }

    private void displayItems(String message, List<PlayListItem> items) {
        StringBuilder sb = new StringBuilder();
        for (PlayListItem item : items) {
            sb.append("\n");
            sb.append("itemId = ");
            sb.append(item.getItemId());
            sb.append("; title = ");
            sb.append(item.getComposition().getTitle());
            sb.append(";");
        }
        String text = sb.toString();
        Log.d("KEK", message + text);
        System.out.println(message + text);
    }
}

File: data/src/androidTest/java/com/github/anrimian/musicplayer/data/database/dao/playlist/PlayListsDaoWrapperTest.kt
package com.github.anrimian.musicplayer.data.database.dao.playlist

import android.util.Log
import androidx.room.Room
import androidx.test.platform.app.InstrumentationRegistry
import com.github.anrimian.musicplayer.data.database.LibraryDatabase
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao
import com.github.anrimian.musicplayer.data.database.entities.playlist.PlaylistEntryEntity
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlaylist
import com.github.anrimian.musicplayer.data.storage.providers.playlists.StoragePlaylistItem
import com.github.anrimian.musicplayer.domain.models.composition.InitialSource
import com.github.anrimian.musicplayer.domain.models.composition.LocalFileStatus
import com.github.anrimian.musicplayer.domain.models.playlist.PlaylistEntry
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class PlayListsDaoWrapperTest {

    private lateinit var playListDao: PlaylistDao
    private lateinit var compositionsDao: CompositionsDao
    private lateinit var db: LibraryDatabase

    private lateinit var daoWrapper: PlaylistsDaoWrapper

    @BeforeEach
    fun setUp() {
        val context = InstrumentationRegistry.getInstrumentation().context
        db = Room.inMemoryDatabaseBuilder(context, LibraryDatabase::class.java).build()
        compositionsDao = db.compositionsDao()
        playListDao = db.playListDao()

        daoWrapper = PlaylistsDaoWrapper(playListDao, compositionsDao, db)
    }

    @AfterEach
    fun tearDown() {
        db.close()
    }

    @Test
    fun testMoveItems() {
        val time = System.currentTimeMillis()
        val playlistId = daoWrapper.insertPlaylist("playlist", time, time) { null }

        for (i in 0..9) {
            val id = DbTestUtils.insert(compositionsDao, null, null, i.toString())
            playListDao.insertPlayListEntries(
                listOf(
                    PlaylistEntryEntity(
                        0,
                        null,
                        id,
                        playlistId,
                        i
                    )
                )
            )
        }

        var items = daoWrapper.getPlayListItemsObservable(playlistId, false, null).blockingFirst()
        displayItems("testMoveItems, items: ", items)

        daoWrapper.moveItems(playlistId, 0, 7)

        items = daoWrapper.getPlayListItemsObservable(playlistId, false, null).blockingFirst()
        displayItems("testMoveItems, moved items: ", items)
    }

    @Test
    fun getNextOrderPositionBasedOnMaxOrderPosition() {
        val time = System.currentTimeMillis()
        val playlistId = daoWrapper.insertPlaylist("playlist", time, time) { null }

        // Sparse positions: 10, 20
        for (i in 0..1) {
            val id = DbTestUtils.insert(compositionsDao, null, null, i.toString())
            playListDao.insertPlayListEntry(null, id, playlistId, if (i == 0) 10 else 20)
        }

        assertEquals(21, daoWrapper.getNextOrderPosition(playlistId))
    }

    @Test
    fun getNextOrderPositionForEmptyPlaylist() {
        val time = System.currentTimeMillis()
        val playlistId = daoWrapper.insertPlaylist("playlist", time, time) { null }

        assertEquals(0, daoWrapper.getNextOrderPosition(playlistId))
    }

    @Test
    fun insertPlayListItemsAppendsAtTheEndUsingGetNextOrderPosition() {
        val time = System.currentTimeMillis()
        val playlistId = daoWrapper.insertPlaylist("playlist", time, time) { null }

        // Start with a sparse position 50
        val id1 = compositionsDao.insert(
            null, null, null, "1", null, null, null, null, "test", 100L, 100L, 1L, 
            time, time, 0, 0, 0, 0, LocalFileStatus.AVAILABLE, null, InitialSource.LOCAL
        )
        playListDao.insertPlayListEntry(null, id1, playlistId, 50)

        // Then append a second item using the default parameter
        val storageId2 = 2L
        compositionsDao.insert(
            null, null, null, "2", null, null, null, null, "test", 100L, 100L, storageId2, 
            time, time, 0, 0, 0, 0, LocalFileStatus.AVAILABLE, null, InitialSource.LOCAL
        )
        val item2 = StoragePlaylistItem(2, storageId2)
        daoWrapper.insertPlayListItems(listOf(item2), playlistId)

        // Check the second item's position, it should be 51
        val items = daoWrapper.getPlayListItemsObservable(playlistId, false, null).blockingFirst()
        assertEquals(2, items.size)
        // items are ordered by orderPosition
        assertEquals(51, playListDao.selectPositionById(items[1].entryId))
    }

    @Test
    fun testUpdatePlaylistNameThatAlreadyExists() {
        val time = System.currentTimeMillis()
        daoWrapper.insertPlaylist("test", time, time) { 1L }
        daoWrapper.insertPlaylist("test1", time, time) { 2L }
        daoWrapper.insertPlaylist("test2", time, time) { 3L }

        val duplicatePlayList2 = StoragePlaylist(2L, "test", time, time)
        val duplicatePlayList3 = StoragePlaylist(3L, "test", time, time)
//        daoWrapper.applyChanges(emptyList(), asList(
//                new Change<>(playList2, duplicatePlayList2),
//                new Change<>(playList3, duplicatePlayList3)
//        ));

        println("KEKAS" + daoWrapper.getPlayListsObservable(null).blockingFirst())
    }

    private fun displayItems(message: String, items: List<PlaylistEntry>) {
        val text = items.joinToString(separator = "\n") { item ->
            "itemId = ${item.entryId}; title = ${item.title};"
        }
        Log.d("KEK", "$message\n$text")
        println("$message\n$text")
    }
}